proposals/0030-markdoc.md

```diff
@@ -0,0 +1,295 @@
+- Start Date: 2023-02-28
+- Reference Issues: https://github.com/withastro/roadmap/issues/496
+- Implementation PR: https://github.com/withastro/astro/pull/6209
+
+# Summary
+
+This RFC introduces Markdoc support to Astro Content Collections.
+
+
+# Example
+
+https://user-images.githubusercontent.com/51384119/218775296-86dd637d-586a-45f1-b374-e0c96be1a567.mp4
+
+
+
+Say you've authored a collection of blog posts using Markdoc. You can store these entries as a collection, identically to Markdown or MDX:
+
+```bash
+src/content/
+	blog/
+		post-1.mdoc
+		post-2.mdoc
+		post-3.mdoc
+...
+```
+
+Then, users can query entry frontmatter with the same `getCollection()` and `getEntryBySlug()` APIs:
+
+```astro
+---
+import { getCollection, getEntryBySlug } from 'astro:content';
+
+const blog = await getCollection('blog');
+const firstEntry = await getEntryBySlug('blog', 'post-1');
+---
+```
+
+Users are also free to render Markdoc contents using a `Content` component, configuring Markdoc tags and component mappings as props:
+
+```astro
+---
+import Title from '../components/Title.astro';
+import Marquee from '../components/Marquee.astro';
+import { getEntryBySlug } from 'astro:content';
+
+const mdocEntry = await getEntryBySlug('blog', 'test');
+const { Content } = await mdocEntry.render();
+---
+
+<html lang="en">
+	<body>
+		<Content
+			config={{
+				variables: { underlineTitle: true },
+			}}
+			components={{
+				h1: Title,
+				marquee: Marquee,
+			}}
+		/>
+	</body>
+</html>
+```
+
+See [the detailed API section](#api) for information on all supported props, shared config, and advanced component mapping.
+
+# Background & Motivation
+
+Include any useful background detail that that explains why this RFC is important.
+What are the problems that this RFC sets out to solve? Why now? Be brief!
+
+It can be useful to illustrate your RFC as a user problem in this section.
+(ex: "Users have reported that it is difficult to do X in Astro today.")
+
+# Goals
+
+- **Create an `@astrojs/markdoc` integration** that adds `.mdoc` support to content collections.
+- **Support Astro components and server-rendered UI components** (React, Vue, Svelte, etc) within Markdoc files. Note this leaves client-rendered UI components out-of-scope (see non-goals).
```

Comment:
I really like that client-side UI component logic is out of scope for now. It feels very Astro to default to HTML-only but then still gives you all of the power of Astro islands inside of Astro's syntax.

Was there any thought to only supporting Astro components and not server-rendered UI components? If it's more complexity to disable non-Astro components than to allow them then I guess I don't feel strongly, but there is a nice, limited-scope MVP quality to Astro-only to start.

Reply:
Yes, it was more engineering work to disable server rendered non-Astro components. We can check for the `AstroComponentFactory` flag to disable this though! I understand this could mean a clearer user story at launch 👍